### DISCUSSION TITLE
Introducing simple-launch

### DISCUSSION THREAD

**Discussion by johnwason** [johnwason/simple-launch#3] [Announcements]
@Levi-Armstrong @marip8 @robinsonmm @marip8 @wolfv @traversaro @Tobias-Fischer @gavanderhoorn

`simple-launch` it a lightweight launch tool (process orchestractor). It has similar goals to roslaunch, but is intended to be a lot simpler and not require ROS. It has the following major benefits:

* Lightweight, single file Python program with minimal dependencies
* Does not require ROS and can run any program necessary
* Various parameters available to control program execution including arguments, working directory, environment, restart, auto quit, etc.
* Better Windows support, using Jobs and Windows Messages to provide graceful shutdown and avoid hanging programs
* A simple GUI that can be displayed to the user with a "Stop All" button, useful for when a launch is done interactively
* A companion "simple-launch-process" library used to catch signals and windows messages, to provide a graceful shutdown signal. The companion libraries are available for Python and C++

https://github.com/johnwason/simple-launch-process
https://github.com/johnwason/simple-launch-process-cpp

**Comment by gavanderhoorn** [comment]
Nice project.

Would it perhaps make sense not to call it `simple-launch`? For anyone not familiar with `roslaunch` -- which you imply is *not* simple -- it doesn't necessarily mean anything this is *simple* launch.

Also: you may end up adding more features which would take it away from its simple roots, at which point the name may no longer be appropriate.

> Better Windows support, using Jobs and Windows Messages to provide graceful shutdown and avoid hanging programs

might be nice to contribute upstream?

**Reply by johnwason** [reply]
I am open to renaming the project. I couldn't think of a better name...

> might be nice to contribute upstream?

`roslaunch` is pretty complicated. If one of the maintainers is interested I can explain how to incorporate better Windows support. Unfortunately it does require some changes to the applications running on Windows because there isn't an equivalent of `signal` in Windows, it requires some work by the programs. The companion libraries implement this functionality here.

**Reply by gavanderhoorn** [reply]
I wasn't necessarily thinking of `roslaunch`. [ros2/launch](https://github.com/ros2/launch) is also supposed to be ROS-agnostic, is also written in Python and supports Windows as well.

> Unfortunately it does require some changes to the applications running on Windows because there isn't an equivalent of `signal` in Windows, it requires some work by the programs. 

hm, that complicates things.

**Reply by johnwason** [reply]
This project is extracted from the PyRI teach pendant launcher (https://github.com/pyri-project/pyri-core) and modified to be standalone. I did look at ros2/launch, but didn't like the way that threading and signaling was handled. This project is designed to be importable as a module, so it may be easier to wrap it than backport the improvements.

I did a short writeup on Windows signaling here: https://github.com/johnwason/simple-launch-process#shutdown-signal-explanation TL;DR windows processes that don't have a main window need to listen for "Windows Messages" and "Console Control Events" to reliably receive a graceful shutdown signal. The companion libraries use a hidden "message only" window to receive the events.

**Reply by johnwason** [reply]
See also https://github.com/johnwason/simple-launch/issues/4 about Linux `cgroup` support. `cgroup` is necessary to make sure that all processes terminate correctly in the event of a crash, but require root access unless specially configured.

**Comment by johnwason** [comment]
I added support for Jinja2 templates and pushed to pypi.org. This should make it easier to work with Conda environments.

**Comment by johnwason** [comment]
What do you guys think of the name `drakkar` or `drekar` instead of simple-launch? It is a type of Viking longboat. Idea came from all rowing together as a team...

**Comment by johnwason** [comment]
Rename to `drekar-launch` is complete
